mlrun/utils/vault.py

```diff
@@ -0,0 +1,269 @@
+
+        response = requests.request(method, full_url, headers=headers, data=data)
+
+        if not response:
+            logger.error(
+                "Vault failed the API call. Response code: ({}) - {}".format(
+                    response.status_code, response.reason
+                )
+            )
+        return response
+
+    # This method logins to the vault, assuming the container has a JWT token mounted as part of its assigned service
+    # account.
+    def _login_with_jwt_token(self, role):
+
+        if role is None:
+            logger.warning(
+                "login_with_token: Role passed is None. Will not attempt login"
+            )
+            return
+
+        jwt_token = self._read_jwt_token()
+
+        login_url = f"{self.url}/v1/auth/kubernetes/login"
+        data = {"jwt": jwt_token, "role": role}
+
+        response = requests.post(login_url, data=json.dumps(data))
+        if not response:
+            logger.error(
+                "login_with_token: Vault failed the login request. Role: {}, Response code: ({}) - {}".format(
+                    role, response.status_code, response.reason
+                )
+            )
+            return
+        self._token = response.json()["auth"]["client_token"]
+
+    def get_secrets(self, keys, user=None, project=None):
+        secret_path = VaultStore._generate_vault_path(user=user, project=project)
+        secrets = {}
+        response = self._vault_api_call("GET", secret_path)
+
+        if not response:
+            return secrets
+
+        values = response.json()["data"]["data"]
+
+        # if no specific keys were asked for, return all the values available
+        if not keys:
+            return values
+
+        for key in keys:
+            if key in values:
+                secrets[key] = values[key]
+        return secrets
+
+    def add_vault_secrets(self, items, project=None, user=None):
+        data_object = {"data": items}
+        url = VaultStore._generate_vault_path(project=project, user=user)
+
+        response = self._vault_api_call("POST", url, data_object)
+        if not response:
+            raise ValueError(
+                f"Vault failed the API call to create secrets. project={project}/user={user}"
+            )
+
+    def delete_vault_secrets(self, project=None, user=None):
+        self._login()
+        # Using the API to delete all versions + metadata of the given secret.
+        url = "v1/secret/metadata/" + VaultStore._generate_vault_path(
+            prefix="", project=project, user=user
+        )
+
+        response = self._vault_api_call("DELETE", url)
+        if not response:
+            raise ValueError(
+                f"Vault failed the API call to delete secrets. project={project}/user={user}"
+            )
+
+    def create_project_policy(self, project):
+        policy_name = "mlrun-project-{}".format(project)
+        # TODO - need to make sure name is escaped properly and invalid chars are stripped
```

Comment:
Please add link to the restriction for policy name

Reply:
Well, it's not really a restriction on policy name - Vault is actually pretty lax on that. The issue is that the policy name is part of the URL path used for the REST call. Therefore, things like '/' or chars that are invalid in a URL will cause this to fail. So it's either escaping the name properly when generating the path or failing if bad chars are part of the name.
Anyway, for now it's good enough without it - the API call will fail if the URL is invalid, and the user will get the error.